### PR TITLE
Fix queue problems

### LINKED ISSUE #438: The entire library is inserted into the queue by selecting a new song
# Steps to reproduce the behavior
- Play any track in your library
- The entire library is inserted in the queue
- By changing the song the full library is inserted back in the queue again, added at the previous queue causing errors like this:

First Screen
Playing a song, the library is added to the queue
It's in italian, "Coda" is "Queue"

Second Screen
Now the full queue is added again, and if you play another song, it is added again and again and it can reach very high numbers like this

Third Screen

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/64861614-the-entire-library-is-inserted-into-the-queue-by-selecting-a-new-song?utm_campaign=plugin&utm_content=tracker%2F60766398&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60766398&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #442: All musics are added to the playlist when you try to added one or a certain set of musics
Hi! Thanks for the awesome work with Elementary OS!
When I try to add a song to a certain playlist, all songs are added to the playlist instead.
I'm adding to the playlist by right-clicking the desired music and "add to the playlist" and choosing the desired one.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65008374-all-musics-are-added-to-the-playlist-when-you-try-to-added-one-or-a-certain-set-of-musics?utm_campaign=plugin&utm_content=tracker%2F60766398&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60766398&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Mohelm97**
Fixes #438 #442 #91 #400 #436

### CODE REVIEW

**Comment by Nine-H** [comment]
I've been running this branch all afternoon and I've noticed that the first track you select is always played twice. This may be out of scope for this patch.

**Comment by Mohelm97** [comment]
It was actually a fast patch that needs a lot of work, that's why I mentioned some, 
but I push it because Music become unusable if you want to change the song (you will have a huge queue),
I will try to work on it more today,
Cheers :D

**Comment by Mohelm97** [comment]
@Nine-H As I see, There's a lot of stuff going wrong here if you noticed we have _current, queue_playlist both should the current playlist, as stated at line 36
TODO: Have a correct implementation based on playlists.
But we have a half implementation of playlists and half on _current so it's a big mess, I don't know if there's actually a difference here or not? Should we just remove one (Obviously _current) or what?

File: src/PlaybackManager.vala [hunk 1]
```diff
@@ -208,27 +211,37 @@ public class Noise.PlaybackManager : Object {
...
+                _current_shuffled.set (0, _current.get(0));
+            }
+            //We don't want to miss one of the songs if current_media == null
+            for (int i = (current_media == null) ? 0 : 1; i < _current.size; i++) {
                 int n_media = temp.size;
                 if (n_media == 0) {
                     break;
                 }
+                // the range for rint_rangeandom [begin..end-1]
+                int random = Random.int_range (0, n_media);
```
Comment (tintou) [inline src/PlaybackManager.vala]:
```suggestion
                int random = Random.int_range (0, n_media - 1);
```

File: src/PlaybackManager.vala [hunk 2]
```diff
@@ -238,18 +251,16 @@ public class Noise.PlaybackManager : Object {
...
-        // next check if user has queued media
-        if (queue_playlist.medias.size > 0) {
-            rv = poll_queue ();
-            _playing_queued_song = true;
-        } else if (main_settings.shuffle_mode != Noise.Settings.Shuffle.OFF) {
-            if (_current_shuffled.is_empty ) {
-                foreach (Media s in library.get_medias ())
-                    add_to_current (s);    //first initialize the current selection the reshuffle it
+        if (main_settings.shuffle_mode != Noise.Settings.Shuffle.OFF) {
+            debug(_current_shuffled.size.to_string());
```
Comment (tintou) [inline src/PlaybackManager.vala]: this line should probably be removed or changed as:
```suggestion
            debug ("Shuffled size: %d", _current_shuffled.size);
```

File: src/PlaybackManager.vala [hunk 3]
```diff
@@ -335,15 +340,17 @@ public class Noise.PlaybackManager : Object {
                 
                 rv = _current.get (_current_index);
             } else {
-                foreach (Media s in library.get_medias ()) {
-                    add_to_current(s);
-                }
-                
+                queue_media (library.get_medias());
```
Comment (tintou) [inline src/PlaybackManager.vala]:
```suggestion
                queue_media (library.get_medias ());
```

File: src/PlaybackManager.vala [hunk 4]
```diff
@@ -358,8 +365,9 @@ public class Noise.PlaybackManager : Object {
         var main_settings = Settings.Main.get_default ();
         if(main_settings.shuffle_mode != Noise.Settings.Shuffle.OFF) {
             if (_current_shuffled.is_empty) {
-                foreach (Media s in library.get_medias ())
-                    add_to_current (s);    //first initialize the current selection the reshuffle it
+                if (_current.is_empty) {
+                    queue_media (library.get_medias ()); // first initialize the current selection the reshuffle it
+                }
```
Comment (tintou) [inline src/PlaybackManager.vala]: empty newline after ending curly bracket

File: src/PlaybackManager.vala [hunk 5]
```diff
@@ -238,18 +251,16 @@ public class Noise.PlaybackManager : Object {
...
-        } else if (main_settings.shuffle_mode != Noise.Settings.Shuffle.OFF) {
-            if (_current_shuffled.is_empty ) {
-                foreach (Media s in library.get_medias ())
-                    add_to_current (s);    //first initialize the current selection the reshuffle it
+        if (main_settings.shuffle_mode != Noise.Settings.Shuffle.OFF) {
+            debug(_current_shuffled.size.to_string());
+            if (_current_shuffled.is_empty) {
+                if (_current.is_empty) {
+                    queue_media (library.get_medias ()); // first initialize the current selection the reshuffle it
+                }
```
Comment (tintou) [inline src/PlaybackManager.vala]: empty newline after ending curly bracket

File: src/PlaybackManager.vala (hunk 1 above)
Comment (tintou) [reply src/PlaybackManager.vala]: indeed, nice catch :)

**Pull request merged**